packages/ec-core/src/operator/selector/recursive_weighted.rs

```diff
@@ -87,154 +130,95 @@ where
     }
 }
 
-pub struct WeightedSelectorPair<P, A, B>
-where
-    P: Population,
-    A: WeightedSelector<P>,
-    B: WeightedSelector<P>,
-{
+pub struct WeightedSelectorPair<A, B> {
     a: A,
     b: B,
-    // weight: usize,
-    _p: PhantomData<P>,
+    distr: Option<Bernoulli>,
+    weight_sum: u32,
 }
 
-pub enum WeightedSelectorsError<P, A, B>
-where
-    P: Population,
-    A: Selector<P>,
-    B: Selector<P>,
-{
-    A(A::Error),
-    B(B::Error),
+#[derive(Debug)]
+pub enum WeightedSelectorsError<A, B> {
+    A(A),
+    B(B),
 
     ZeroWeightSum,
 
     // #[error("Adding the weights {0} and {1} overflows")]
-    WeightSumOverflows(usize, usize),
+    WeightSumOverflows(u32, u32),
 }
 
-impl<P, A, B> std::fmt::Debug for WeightedSelectorsError<P, A, B>
+impl<P, A, B> Selector<P> for WeightedSelectorPair<A, B>
 where
     P: Population,
-    A: Selector<P>,
-    A::Error: std::fmt::Debug,
-    B: Selector<P>,
-    B::Error: std::fmt::Debug,
+    A: WeightedSelector + Selector<P>,
+    B: WeightedSelector + Selector<P>,
 {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            Self::A(arg0) => f.debug_tuple("A").field(arg0).finish(),
-            Self::B(arg0) => f.debug_tuple("B").field(arg0).finish(),
-            Self::ZeroWeightSum => write!(f, "ZeroWeightSum"),
-            Self::WeightSumOverflows(arg0, arg1) => f
-                .debug_tuple("WeightSumOverflows")
-                .field(arg0)
-                .field(arg1)
-                .finish(),
-        }
-    }
-}
-
-impl<P, A, B> Selector<P> for WeightedSelectorPair<P, A, B>
-where
-    P: Population,
-    A: WeightedSelector<P>,
-    B: WeightedSelector<P>,
-{
-    type Error = WeightedSelectorsError<P, A, B>;
+    type Error = WeightedSelectorsError<A::Error, B::Error>;
 
     fn select<'pop>(
         &self,
         population: &'pop P,
         rng: &mut rand::prelude::ThreadRng,
     ) -> Result<&'pop <P as Population>::Individual, Self::Error> {
-        let a_weight = self.a.weight();
-        let b_weight = self.b.weight();
-        // For performance reasons, it would be nice to check that the sum works
-        // and is > 0 at construction time so we don't have to do it here.
-        let weight_sum = a_weight
-            .checked_add(b_weight)
-            .ok_or_else(|| WeightedSelectorsError::WeightSumOverflows(a_weight, b_weight))?;
-        if weight_sum == 0 {
+        let Some(distr) = self.distr else {
```

Comment:
Doesn't this want to return something different in the `else` case?

Reply:
The distr is only not set in the constructor in the zero sum case tho so that error makes sense here imho.